ctru-rs/src/services/ndsp/mod.rs

```diff
@@ -0,0 +1,223 @@
+pub mod wave;
+use wave::{WaveInfo, WaveStatus};
+
+use crate::error::ResultCode;
+use crate::services::ServiceReference;
+
+use std::sync::Mutex;
+
+#[derive(Copy, Clone, Debug)]
+#[repr(u32)]
+pub enum OutputMode {
+    Mono = ctru_sys::NDSP_OUTPUT_MONO,
+    Stereo = ctru_sys::NDSP_OUTPUT_STEREO,
+    Surround = ctru_sys::NDSP_OUTPUT_SURROUND,
+}
+
+#[derive(Copy, Clone, Debug)]
+#[repr(u32)]
+pub enum AudioFormat {
+    PCM8Mono = ctru_sys::NDSP_FORMAT_MONO_PCM8,
+    PCM16Mono = ctru_sys::NDSP_FORMAT_MONO_PCM16,
+    ADPCMMono = ctru_sys::NDSP_FORMAT_MONO_ADPCM,
+    PCM8Stereo = ctru_sys::NDSP_FORMAT_STEREO_PCM8,
+    PCM16Stereo = ctru_sys::NDSP_FORMAT_STEREO_PCM16,
+}
+
+#[derive(Copy, Clone, Debug)]
+#[repr(u32)]
+pub enum InterpolationType {
+    Polyphase = ctru_sys::NDSP_INTERP_POLYPHASE,
+    Linear = ctru_sys::NDSP_INTERP_LINEAR,
+    None = ctru_sys::NDSP_INTERP_NONE,
+}
+
+pub struct Channel(i32);
+
+static NDSP_ACTIVE: Mutex<usize> = Mutex::new(0);
+
+pub struct Ndsp {
+    _service_handler: ServiceReference,
+}
+
+impl Ndsp {
+    pub fn init() -> crate::Result<Self> {
+        let _service_handler = ServiceReference::new(
+            &NDSP_ACTIVE,
+            false,
+            || {
+                ResultCode(unsafe { ctru_sys::ndspInit() })?;
+
+                Ok(())
+            },
+            || unsafe {
+                ctru_sys::ndspExit();
+            },
+        )?;
+
+        Ok(Self { _service_handler })
+    }
+
+    /// Return the representation of the specified channel.
+    ///
+    /// # Errors
+    ///
+    /// An error will be returned if the channel id is not between 0 and 23.
+    pub fn channel(&self, id: u8) -> crate::Result<Channel> {
+        if id > 23 {
+            return Err(crate::Error::InvalidChannel(id.into()));
+        }
+
+        Ok(Channel(id.into()))
+    }
+
+    /// Set the audio output mode. Defaults to `OutputMode::Stereo`.
+    pub fn set_output_mode(&mut self, mode: OutputMode) {
+        unsafe { ctru_sys::ndspSetOutputMode(mode as u32) };
+    }
+}
+
+// All channel operations are thread-safe thanks to `libctru`'s use of thread locks.
+// As such, there is no need to hold channels to ensure correct mutability.
+// As such, this struct is more of a dummy than an actually functional block.
+impl Channel {
+    /// Reset the channel
+    pub fn reset(&self) {
+        unsafe { ctru_sys::ndspChnReset(self.0) };
+    }
+
+    /// Initialize the channel's parameters
+    pub fn init_parameters(&self) {
```

Comment:
> You mentioned that all those operations are thread-safe though so maybe just `init` and cleanup would be enough, although it might be weird if a program got two instances of the same channel and it ended up resetting while some other thing was using it...

That's exactly why I did it this way. The other option would be to have references to the `Ndsp` object and to count for `Channel` references to avoid multiple instances.

I believe the reference is needed either way (right now `Channels` can outlive `Ndsp`, bad behaviour), but I was wondering if there were better ways to count channels than just using a 24bit flag...

Reply:
Here's a possible kind of implementation? https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e136b260313bbca1e62f31dc4fddaf99

It allows aliasing on the channels with immutable refs, only initializes / resets each one once (although perhaps manually resetting would be trickyier), and does so lazily as-needed. It's probably not perfect and might be missing some use cases but it at least ties the lifetimes together.

Reply:
> Maybe replace the `Cell` with an `AtomicBool` and use `swap(true, Ordering::SeqCst)` to avoid two threads both racing to initialize the channel (even though libctru makes it thread safe, we probably don't want to initialize twice and possibly lose one thread's settings).

That can't happen (I think?). Our `Ndsp` struct is blocked by our own mutex which locks it to only one thread.

Reply:
The `Cell` approach wasn’t doable, but I managed to “merge” the lifetime check and the exclusivity check by using `RefCell` in `Ndsp`. As always, it’s a bit heavy on the system, but it’s needed for safety…